File: typescript/tests/security.test.ts
import { describe, it, expect } from 'vitest';
import {
  timingSafeEqual, validateDid, validateHash, validateSignature,
  validateTimestamp, validateVersion, validateTrustTier, validateAttestationCount,
  keccak256, snapshotHash, ed25519Sign, ed25519Verify, generateKeyPair,
  canonicalizeJson, MerkleTree,
} from '../src/index.js';
import { AgentPassport } from '../src/passport.js';
import { WorkReceipt } from '../src/receipt.js';
import type { PassportConfig, Skill, Soul, Policies, Lineage } from '../src/types.js';

function testConfig(): PassportConfig {
  return {
    id: 'did:key:z6MkpTHR8VNsBxYAAWHut2Geadd9jSwuBV8xRoAnwWsdvktH',
    publicKey: 'z6MkpTHR8VNsBxYAAWHut2Geadd9jSwuBV8xRoAnwWsdvktH',
    ownerDID: 'did:key:z6MkpOwnerABCDEFGHIJKLMNOPQRSTUVWXYZ12345',
    skills: [{
      name: 'test', version: '1.0.0', description: 'Test',
      capabilities: ['test'], hash: '0x' + '00'.repeat(32),
    }],
    soul: {
      personality: 'focused', work_style: 'test', constraints: ['none'],
      hash: '0x' + '00'.repeat(32), frozen: false,
    },
    policies: { policy_set_hash: '0x' + '00'.repeat(32), summary: ['can_bid'] },
    lineage: { kind: 'single', parents: [], generation: 0 },
  };
}

// --- Timing-Safe Comparison ---
describe('timingSafeEqual', () => {
  it('returns true for equal strings', () => {
    expect(timingSafeEqual('abc', 'abc')).toBe(true);
  });
  it('returns false for different strings', () => {
    expect(timingSafeEqual('abc', 'abd')).toBe(false);
  });
  it('returns false for different lengths', () => {
    expect(timingSafeEqual('short', 'longer')).toBe(false);
  });
});

// --- Validation ---
describe('validation', () => {
  it('validates valid DID', () => {
    expect(() => validateDid('did:key:z6MkpTHR8VNsBxYAAWHut2Geadd9jSwuBV8xRoAnwWsdvktH')).not.toThrow();
  });
  it('rejects empty DID', () => {
    expect(() => validateDid('')).toThrow();
  });
  it('rejects invalid DID', () => {
    expect(() => validateDid('not-a-did')).toThrow();
    expect(() => validateDid('did:key:abc')).toThrow();
  });
  it('validates valid hash', () => {
    expect(() => validateHash('0x' + 'ab'.repeat(32))).not.toThrow();
  });
  it('rejects invalid hash', () => {
    expect(() => validateHash('')).toThrow();
    expect(() => validateHash('0x123')).toThrow();
  });
  it('validates valid signature', () => {
    expect(() => validateSignature('ab'.repeat(64))).not.toThrow();
  });
  it('rejects invalid signature', () => {
    expect(() => validateSignature('')).toThrow();
    expect(() => validateSignature('short')).toThrow();
  });
  it('validates valid timestamp', () => {
    expect(() => validateTimestamp('2026-02-15T12:00:00Z')).not.toThrow();
  });
  it('rejects invalid timestamp', () => {
    expect(() => validateTimestamp('')).toThrow();
    expect(() => validateTimestamp('not-a-date')).toThrow();
  });
  it('validates version', () => {
    expect(() => validateVersion(1)).not.toThrow();
    expect(() => validateVersion(0)).toThrow();
    expect(() => validateVersion(-1)).toThrow();
  });
  it('validates trust tier', () => {
    expect(() => validateTrustTier(0)).not.toThrow();
    expect(() => validateTrustTier(3)).not.toThrow();
    expect(() => validateTrustTier(-1)).toThrow();
    expect(() => validateTrustTier(4)).toThrow();
    expect(() => validateTrustTier(999)).toThrow();
  });
  it('validates attestation count', () => {
    expect(() => validateAttestationCount(0)).not.toThrow();
    expect(() => validateAttestationCount(-1)).toThrow();
  });
});

// --- Signature Forgery ---
describe('signature forgery', () => {
  it('tampered passport fails verification', async () => {
    const { publicKey, privateKey } = generateKeyPair();
    const p = await AgentPassport.create(testConfig());
    await p.sign(privateKey);
    p.data.snapshot.version = 999;
    expect(await p.verify(publicKey)).toBe(false);
  });
});

// --- Hash Manipulation ---
describe('hash manipulation', () => {
  it('different content produces different hashes', () => {
    const h1 = snapshotHash({ skill: 'go', version: 1 });
    const h2 = snapshotHash({ skill: 'go', version: 2 });
    expect(h1).not.toBe(h2);
  });
});

// --- Replay Attack ---
describe('replay attack', () => {
  it('stolen proof fails on different passport', async () => {
    const { publicKey, privateKey } = generateKeyPair();
    const p1 = await AgentPassport.create(testConfig());
    await p1.sign(privateKey);
    const stolenProof = p1.data.proof;

    const cfg2 = testConfig();
    cfg2.id = 'did:key:z6MkDIFFERENT1234567890ABCDEFGHIJKLMNOPQRSTUVWXYZ';
    const p2 = await AgentPassport.create(cfg2);
    p2.data.proof = stolenProof;
    expect(await p2.verify(publicKey)).toBe(false);
  });
});

// --- Null/Empty Injection ---
describe('null/empty injection', () => {
  it('rejects empty id', async () => {
    const cfg = testConfig();
    cfg.id = '';
    await expect(AgentPassport.create(cfg)).rejects.toThrow();
  });
  it('rejects empty publicKey', async () => {
    const cfg = testConfig();
    cfg.publicKey = '';
    await expect(AgentPassport.create(cfg)).rejects.toThrow();
  });
  it('rejects empty ownerDID', async () => {
    const cfg = testConfig();
    cfg.ownerDID = '';
    await expect(AgentPassport.create(cfg)).rejects.toThrow();
  });
});

// --- Oversized Input ---
describe('oversized input', () => {
  it('handles 10K skills without crash', async () => {
    const skills: Skill[] = [];
    for (let i = 0; i < 10000; i++) {
      skills.push({
        name: `skill-${i}`, version: '1.0.0', description: 'x',
        capabilities: ['test'], hash: '0x' + '00'.repeat(32),
      });
    }
    const cfg = testConfig();
    cfg.skills = skills;
    const p = await AgentPassport.create(cfg);
    expect(p.data.snapshot.hash).toBeTruthy();
  });
});

// --- Unicode Edge Cases ---
describe('unicode edge cases', () => {
  it('emoji in skills', () => {
    expect(snapshotHash({ skill: '🤖' })).toMatch(/^0x/);
  });
  it('RTL text', () => {
    expect(snapshotHash({ skill: 'مرحبا' })).toMatch(/^0x/);
  });
  it('null bytes', () => {
    expect(snapshotHash({ skill: 'a\x00b' })).toMatch(/^0x/);
  });
});

// --- Integer Overflow ---
describe('integer overflow', () => {
  it('rejects trust tier 999', () => {
    expect(() => validateTrustTier(999)).toThrow();
  });
  it('rejects negative attestation', () => {
    expect(() => validateAttestationCount(-1)).toThrow();
  });
});

// --- Proof Stripping ---
describe('proof stripping', () => {
  it('verify without proof throws', async () => {
    const { publicKey } = generateKeyPair();
    const p = await AgentPassport.create(testConfig());
    await expect(p.verify(publicKey)).rejects.toThrow('no proof present');
  });
});

// --- Key Mismatch ---
describe('key mismatch', () => {
  it('wrong key returns false', async () => {
    const { privateKey } = generateKeyPair();
    const { publicKey: pub2 } = generateKeyPair();
    const p = await AgentPassport.create(testConfig());
    await p.sign(privateKey);
    expect(await p.verify(pub2)).toBe(false);
  });
});

// --- Frozen Mutation ---
describe('frozen mutation', () => {
  it('addSkill throws when frozen', async () => {
    const p = await AgentPassport.create(testConfig());
    p.data.snapshot.skills.frozen = true;
    expect(() => p.addSkill({
      name: 'new', version: '1.0.0', description: 'x',
      capabilities: ['test'], hash: '0x' + '00'.repeat(32),
    })).toThrow('skills are frozen');
  });
});

// --- Canonical JSON Edge Cases ---
describe('canonical JSON edge cases', () => {
  it('null/true/false', () => {
    expect(canonicalizeJson({ a: null, b: true, c: false }))
      .toBe('{"a":null,"b":true,"c":false}');
  });
  it('nested unsorted keys', () => {
    expect(canonicalizeJson({ z: { b: 2, a: 1 }, a: 'first' }))
      .toBe('{"a":"first","z":{"a":1,"b":2}}');
  });
  it('mixed array', () => {
    expect(canonicalizeJson([1, 'two', true, null, { b: 2, a: 1 }]))
      .toBe('[1,"two",true,null,{"a":1,"b":2}]');
  });
  it('special chars', () => {
    const result = canonicalizeJson({ quote: 'a"b', backslash: 'a\\b', newline: 'a\nb' });
    expect(result).toContain('"a\\"b"');
    expect(result).toContain('"a\\\\b"');
    expect(result).toContain('"a\\nb"');
  });
  it('numbers', () => {
    const result = canonicalizeJson({ zero: 0, one: 1, neg: -1, float: 1.5 });
    expect(result).toContain('"float":1.5');
    expect(result).toContain('"neg":-1');
  });
  it('emoji string', () => {
    const result = canonicalizeJson({ emoji: '😀' });
    expect(result).toContain('😀');
  });
  it('deterministic', () => {
    const obj = { z: 1, a: 2, m: { c: 3, b: 4 } };
    expect(canonicalizeJson(obj)).toBe(canonicalizeJson(obj));
  });
});

// --- Merkle Timing-Safe ---
describe('merkle timing-safe', () => {
  it('rejects wrong root', () => {
    const enc = new TextEncoder();
    const leaves = [0, 1, 2, 3].map(i => keccak256(enc.encode(`leaf-${i}`)));
    const tree = new MerkleTree(leaves);
    const root = tree.root();
    const proof = tree.proof(0);
    expect(MerkleTree.verifyProof(leaves[0], root, proof, 0)).toBe(true);
    const fakeRoot = '0x' + '00'.repeat(32);
    expect(MerkleTree.verifyProof(leaves[0], fakeRoot, proof, 0)).toBe(false);
  });
});

File: go/crypto/security_test.go
package crypto

import (
	"fmt"
	"math"
	"strings"
	"testing"
)

// --- Timing-Safe Comparison Tests ---

func TestTimingSafeEqual_Match(t *testing.T) {
	a := "0xabcdef1234567890abcdef1234567890abcdef1234567890abcdef1234567890"
	if !TimingSafeEqual(a, a) {
		t.Error("identical strings should be equal")
	}
}

func TestTimingSafeEqual_Mismatch(t *testing.T) {
	a := "0xabcdef1234567890abcdef1234567890abcdef1234567890abcdef1234567890"
	b := "0xabcdef1234567890abcdef1234567890abcdef1234567890abcdef1234567891"
	if TimingSafeEqual(a, b) {
		t.Error("different strings should not be equal")
	}
}

func TestTimingSafeEqual_DifferentLengths(t *testing.T) {
	if TimingSafeEqual("short", "longer string") {
		t.Error("different length strings should not be equal")
	}
}

// --- Validation Tests ---

func TestValidateDID_Valid(t *testing.T) {
	if err := ValidateDID("did:key:z6MkpTHR8VNsBxYAAWHut2Geadd9jSwuBV8xRoAnwWsdvktH"); err != nil {
		t.Errorf("valid DID rejected: %v", err)
	}
}

func TestValidateDID_Empty(t *testing.T) {
	if err := ValidateDID(""); err == nil {
		t.Error("empty DID should fail")
	}
}

func TestValidateDID_Invalid(t *testing.T) {
	invalids := []string{"not-a-did", "did:key:abc", "did:web:example.com", "did:key:z6Mk"}
	for _, d := range invalids {
		if err := ValidateDID(d); err == nil {
			t.Errorf("invalid DID %q should fail", d)
		}
	}
}

func TestValidateHash_Valid(t *testing.T) {
	if err := ValidateHash("0xabcdef1234567890abcdef1234567890abcdef1234567890abcdef1234567890"); err != nil {
		t.Errorf("valid hash rejected: %v", err)
	}
}

func TestValidateHash_Invalid(t *testing.T) {
	invalids := []string{"", "abcdef", "0x123", "0xGGGG567890abcdef1234567890abcdef1234567890abcdef1234567890abcdef"}
	for _, h := range invalids {
		if err := ValidateHash(h); err == nil {
			t.Errorf("invalid hash %q should fail", h)
		}
	}
}

func TestValidateSignature_Valid(t *testing.T) {
	sig := strings.Repeat("ab", 64)
	if err := ValidateSignature(sig); err != nil {
		t.Errorf("valid sig rejected: %v", err)
	}
}

func TestValidateSignature_Invalid(t *testing.T) {
	if err := ValidateSignature(""); err == nil {
		t.Error("empty sig should fail")
	}
	if err := ValidateSignature("tooshort"); err == nil {
		t.Error("short sig should fail")
	}
}

func TestValidateTimestamp_Valid(t *testing.T) {
	if err := ValidateTimestamp("2026-02-15T12:00:00Z"); err != nil {
		t.Errorf("valid timestamp rejected: %v", err)
	}
}

func TestValidateTimestamp_Invalid(t *testing.T) {
	if err := ValidateTimestamp(""); err == nil {
		t.Error("empty timestamp should fail")
	}
	if err := ValidateTimestamp("not-a-date"); err == nil {
		t.Error("invalid timestamp should fail")
	}
}

func TestValidateVersion_Valid(t *testing.T) {
	if err := ValidateVersion(1); err != nil {
		t.Error("version 1 should be valid")
	}
}

func TestValidateVersion_Invalid(t *testing.T) {
	if err := ValidateVersion(0); err == nil {
		t.Error("version 0 should fail")
	}
	if err := ValidateVersion(-1); err == nil {
		t.Error("negative version should fail")
	}
}

func TestValidateTrustTier_Valid(t *testing.T) {
	for i := 0; i <= 3; i++ {
		if err := ValidateTrustTier(i); err != nil {
			t.Errorf("tier %d should be valid", i)
		}
	}
}

func TestValidateTrustTier_Invalid(t *testing.T) {
	if err := ValidateTrustTier(-1); err == nil {
		t.Error("tier -1 should fail")
	}
	if err := ValidateTrustTier(4); err == nil {
		t.Error("tier 4 should fail")
	}
	if err := ValidateTrustTier(999); err == nil {
		t.Error("tier 999 should fail")
	}
}

func TestValidateAttestationCount_Invalid(t *testing.T) {
	if err := ValidateAttestationCount(-1); err == nil {
		t.Error("negative attestation count should fail")
	}
}

// --- Signature Forgery ---

func TestSignatureForgery(t *testing.T) {
	pub, priv, _ := GenerateKeyPair()
	data := []byte("original passport data")
	sig := Ed25519Sign(priv, data)

	// Tamper with data
	tampered := []byte("tampered passport data")
	ok, _ := Ed25519Verify(pub, tampered, sig)
	if ok {
		t.Error("tampered data should not verify")
	}
}

// --- Hash Manipulation ---

func TestHashManipulation(t *testing.T) {
	payload1 := map[string]interface{}{"skill": "go", "version": 1}
	payload2 := map[string]interface{}{"skill": "go", "version": 2}
	h1, _ := SnapshotHash(payload1)
	h2, _ := SnapshotHash(payload2)
	if h1 == h2 {
		t.Error("different payloads should produce different hashes")
	}
}

// --- Replay Attack ---

func TestReplayAttack(t *testing.T) {
	pub, priv, _ := GenerateKeyPair()
	data1 := []byte("passport-1-data")
	sig := Ed25519Sign(priv, data1)

	data2 := []byte("passport-2-data")
	ok, _ := Ed25519Verify(pub, data2, sig)
	if ok {
		t.Error("signature from one document should not verify on another")
	}
	// Same key, same sig, correct data should still work
	ok, _ = Ed25519Verify(pub, data1, sig)
	if !ok {
		t.Error("original should still verify")
	}
}

// --- Key Mismatch ---

func TestKeyMismatch(t *testing.T) {
	_, privA, _ := GenerateKeyPair()
	pubB, _, _ := GenerateKeyPair()
	data := []byte("test data")
	sig := Ed25519Sign(privA, data)
	ok, _ := Ed25519Verify(pubB, data, sig)
	if ok {
		t.Error("wrong key should not verify")
	}
}

// --- Oversized Input ---

func TestOversizedInput(t *testing.T) {
	skills := make([]interface{}, 10000)
	for i := range skills {
		skills[i] = fmt.Sprintf("skill-%d", i)
	}
	payload := map[string]interface{}{"skills": skills}
	h, err := SnapshotHash(payload)
	if err != nil {
		t.Fatalf("oversized input should not crash: %v", err)
	}
	if h == "" {
		t.Error("hash should not be empty")
	}
}

// --- Unicode Edge Cases ---

func TestUnicodeEdgeCases(t *testing.T) {
	cases := []map[string]interface{}{
		{"skill": "🤖"},
		{"skill": "مرحبا"},
		{"skill": "a\x00b"},
		{"skill": "emoji 🎯 test"},
	}
	for _, c := range cases {
		h, err := SnapshotHash(c)
		if err != nil {
			t.Fatalf("unicode case failed: %v", err)
		}
		if h == "" {
			t.Error("hash should not be empty for unicode input")
		}
	}
}

// --- Integer Overflow ---

func TestIntegerOverflow(t *testing.T) {
	if err := ValidateVersion(math.MaxInt64); err != nil {
		t.Error("max int version should be valid")
	}
	if err := ValidateTrustTier(999); err == nil {
		t.Error("tier 999 should be invalid")
	}
	if err := ValidateAttestationCount(-1); err == nil {
		t.Error("negative attestation should be invalid")
	}
}

// --- Canonical JSON Edge Cases ---

func TestCanonicalJSON_NullTrueFalse(t *testing.T) {
	input := map[string]interface{}{
		"a": nil,
		"b": true,
		"c": false,
	}
	result, err := CanonicalizeJSON(input)
	if err != nil {
		t.Fatal(err)
	}
	expected := `{"a":null,"b":true,"c":false}`
	if string(result) != expected {
		t.Errorf("got %s, want %s", string(result), expected)
	}
}

func TestCanonicalJSON_NestedUnsortedKeys(t *testing.T) {
	input := map[string]interface{}{
		"z": map[string]interface{}{
			"b": 2,
			"a": 1,
		},
		"a": "first",
	}
	result, err := CanonicalizeJSON(input)
	if err != nil {
		t.Fatal(err)
	}
	expected := `{"a":"first","z":{"a":1,"b":2}}`
	if string(result) != expected {
		t.Errorf("got %s, want %s", string(result), expected)
	}
}

func TestCanonicalJSON_MixedArray(t *testing.T) {
	input := []interface{}{1, "two", true, nil, map[string]interface{}{"b": 2, "a": 1}}
	result, err := CanonicalizeJSON(input)
	if err != nil {
		t.Fatal(err)
	}
	expected := `[1,"two",true,null,{"a":1,"b":2}]`
	if string(result) != expected {
		t.Errorf("got %s, want %s", string(result), expected)
	}
}

func TestCanonicalJSON_SpecialChars(t *testing.T) {
	input := map[string]interface{}{
		"quote":     `a"b`,
		"backslash": `a\b`,
		"newline":   "a\nb",
	}
	result, err := CanonicalizeJSON(input)
	if err != nil {
		t.Fatal(err)
	}
	// Just verify it doesn't crash and is deterministic
	result2, _ := CanonicalizeJSON(input)
	if string(result) != string(result2) {
		t.Error("not deterministic for special chars")
	}
}

func TestCanonicalJSON_Numbers(t *testing.T) {
	input := map[string]interface{}{
		"zero":     0,
		"one":      1,
		"negative": -1,
		"float":    1.5,
	}
	result, err := CanonicalizeJSON(input)
	if err != nil {
		t.Fatal(err)
	}
	if string(result) == "" {
		t.Error("should not be empty")
	}
}

// --- Merkle Proof with Timing-Safe Comparison ---

func TestMerkleVerifyProof_TimingSafe(t *testing.T) {
	leaves := make([]string, 4)
	for i := range leaves {
		leaves[i] = Keccak256([]byte(fmt.Sprintf("leaf-%d", i)))
	}
	tree := NewMerkleTree(leaves)
	root := tree.Root()
	proof := tree.Proof(0)

	// Correct verification
	if !VerifyProof(leaves[0], root, proof, 0) {
		t.Error("valid proof should verify")
	}

	// Wrong root (timing-safe comparison should still reject)
	fakeRoot := "0x0000000000000000000000000000000000000000000000000000000000000000"
	if VerifyProof(leaves[0], fakeRoot, proof, 0) {
		t.Error("wrong root should not verify")
	}
}
